### ISSUE TITLE
Stack Overflow at ExpressionBasedModel#presolve(ExpressionsBasedModel)

### ISSUE THREAD

**Issue by younes-abouelnagah** [optimatika/ojAlgo#7]
Hi,

First of all, thanks for open sourcing this promising library. I am evaluating it against JOptimise and I found that the current implementation of the presolve function in ExpressionsBasedModel causes a stack overflow because of the recursive call on line 125 (version 38). 

I suggest doing 3 things:

1) use this code instead:
```
    /**
     * Avoids stack overflow exceptions caused by the original  {@link ExpressionsBasedModel#presolve()}.
     *
     * @param model
     */
    public static final void presolveIterative(final ExpressionsBasedModel model) {

        int iters = 0;
        boolean stillSimplifying = true;
        while (stillSimplifying) {
            ++iters;
            if (iters % 100 == 0) {
                log.info("Done {} iterations of presolving");
            }

            stillSimplifying = false;

            for (final Expression tmpExpression : model.getExpressions()) {
                if (tmpExpression.simplify()) {
                    log.debug("Following expression is now redundant: {}", tmpExpression);
                    stillSimplifying = true;
                    break; // Restart the process after removing something.
                }
            }
        }
    }
```

2) Make the presolve functions normal methods.. you have one presolve function as final and the other as static, which prevents subclasses from overriding either. This makes it unnecessarily harder to extend your library, and yet not impossible.

3) Create a tag on github for each release, so that anybody can fork the version they are using and submit a bug fix pull request. Right now I would need to pull master and fix the bug in a version that is unstable and not ready for release. 

Thanks again.

**Comment by apete** [comment]
Thanks for the feedback!

1) Perhaps an iterative solution is better than a recursive, but how deep did you recurse? Do you have a test case that causes a stack overflow? There must be some other, more direct, problem.

2) ExpressionsBasedModel, Expression and Variable are all final... Which part would you like to extend?

3) Yes. (All major versions are tagged in the CVS repository at SourceForge - that is still used in parallel.) Not sure why I haven't done the same at GitHub. Will do in the future.

**Comment by younes-abouelnagah** [comment]
Hi Anders,

1) I don't have a test case, but I have a real problem that I am using for evaluation. It is a knap-sack constrained problem on a supply-demand graph with hundreds of supply nodes (thus thousands of edges). The code I sent you finishes simplifying successfully, removing a bunch of constraints. However, when I invoke the minimize method, it calls the original presolve method and fails with a Stack Oveflow.

2) I wanted to replace the presolve method.

3) Thanks.. looking forward for that.

**Comment by apete** [comment]
Is there nothing you can send me that demonstrates the problem?

**Comment by younes-abouelnagah** [comment]
There were multiple issues leading to the stack overflow exception. There was underlying problem in the model creation code where I set the lower limit to 0 instead of the actual lower limit, and with that out of the way I am able to use OjAlgo much more effectively. However, there is still a problem with the use of recursion while iterative over a collection that can be arbitrarily large, which causes the stack overflow. This can be solved by adopting an iterative solution. I can do that if you create a tag for Version 38 which I can fork.

**Comment by younes-abouelnagah** [comment]
Here is an example of the stack trace I get. Please notice that the algorithm converges if I use a smaller subset of the problem, and I don't see any reason why it wouldn't converge with a larger problem.. if only it doesn't stumble on the Stack Overflow Error.

```
ojAlgo includes a small set of predefined hardware profiles,
none of which were deemed suitable for the hardware you're currently using.
You should set org.ojalgo.OjAlgoUtils.ENVIRONMENT to something that matches the hardware/OS/JVM you're running on.
Additionally it would be appreciated if you contribute your hardware profile to ojAlgo.
https://lists.sourceforge.net/lists/listinfo/ojalgo-user
Exception in thread "main" java.lang.StackOverflowError
	at java.math.MutableBigInteger.divideKnuth(MutableBigInteger.java:1203)
	at java.math.MutableBigInteger.divideKnuth(MutableBigInteger.java:1163)
	at java.math.BigInteger.divideAndRemainderKnuth(BigInteger.java:2095)
	at java.math.BigInteger.divideAndRemainder(BigInteger.java:2083)
	at java.math.BigDecimal.createAndStripZerosToMatchScale(BigDecimal.java:4373)
	at java.math.BigDecimal.divideAndRound(BigDecimal.java:4263)
	at java.math.BigDecimal.divide(BigDecimal.java:4674)
	at java.math.BigDecimal.divide(BigDecimal.java:1756)
	at org.ojalgo.function.BigFunction$13.invoke(BigFunction.java:173)
	at org.ojalgo.function.BigFunction$13.invoke(BigFunction.java:1)
	at org.ojalgo.optimisation.Expression.simplify(Expression.java:869)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:124)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
	at org.ojalgo.optimisation.ExpressionsBasedModel.presolve(ExpressionsBasedModel.java:125)
```

**Comment by apete** [comment]
If you're using v38.x you really should be using v38.2. I have created a tag for v38.2 (the presolve functionality has been improved) and there shouldn't be any problem evaluating the optimisation parts of ojAlgo using the current/master/head.

If you want to submit a pull request for an iterative presolver, then go ahead. I have been thinking about refactoring the presolve implementation so that users can turn it off and/or add different presolve modules. When I do this I'll make sure it is iterative rather than recursive...

**Comment by younes-abouelnagah** [comment]
That's great Adrian. Thank you very much. I will pull both V38.2 and master to make the changes to them. I am also open to discussions about Object Oriented designs that allow for plugging in different algorithms for different parts of code. My first pointer would be the Strategy Pattern (http://www.oodesign.com/strategy-pattern.html).. it is pretty basic but gets the job done!

**Comment by younes-abouelnagah** [comment]
Submitted one pull request for the strategy suggestion: https://github.com/optimatika/ojAlgo/pull/8

**Comment by younes-abouelnagah** [comment]
Submitted another pull request that only changes the recursive static function to an iterative static function: https://github.com/optimatika/ojAlgo/pull/9 

This pull request is created against master even though it should be created against a v38 branch. However, there is no such branch as v38.2 is only a tag. Please watch out if you merge it.

**Comment by apete** [comment]
Thanks for your contribution. I accepted that pull request, and I assume it solved your problem. Therefore I will close this issue.

This is a general improvement that should be part of the next release (all future releases) - so why not put it in the master branch?

I do NOT maintain multiple versions/branches simultaneously as it is too much work. I did however create a Java7 to enable important/major bug fixing without upgrading to Java 8. I dod not work in that branch at all.
